### PR TITLE
Update installation instructions

### PR DESCRIPTION
**Pull request by themperek**
Related to #1289 

Help needed!

### CODE REVIEW

File: documentation/source/quickstart.rst [hunk 1]
```diff
@@ -18,104 +18,83 @@ Cocotb has the following requirements:
...
 .. code-block:: bash
 
-    sudo apt-get install git make gcc g++ swig python-dev
+    brew install python icarus-verilog gtkwave
 
-Red Hat-based
-^^^^^^^^^^^^^
-
-.. code-block:: bash
+For more installation options (also 32-bit simulator) please see `our Wiki <https://github.com/cocotb/cocotb/wiki/Tier-2-Setup-Instructions>`_.
```
Comment (cmarqu) [inline documentation/source/quickstart.rst:62]:
```suggestion
For more installation options (also for 32-bit simulators) please see `our Wiki <https://github.com/cocotb/cocotb/wiki/Tier-2-Setup-Instructions>`_.
```

File: documentation/source/quickstart.rst [hunk 2]
```diff
@@ -18,104 +18,83 @@ Cocotb has the following requirements:
...
 
-    sudo yum install gcc gcc-c++ libstdc++-devel swig python-devel
+Simulator Installation
+^^^^^^^^^^^^^^^^^^^^^^
 
+For details and the list of supported simulators see :ref:`_simulator-support`.
 
-Windows Installation
---------------------
+Install and ensure that simulator executable is in the ``PATH``.
```
Comment (cmarqu) [inline documentation/source/quickstart.rst:69]:
```suggestion
Install and ensure that the simulator executable is on the ``PATH``.   
```

File: documentation/source/quickstart.rst [hunk 3]
```diff
@@ -18,104 +18,83 @@ Cocotb has the following requirements:
...
-    sudo yum install gcc gcc-c++ libstdc++-devel swig python-devel
+Simulator Installation
+^^^^^^^^^^^^^^^^^^^^^^
 
+For details and the list of supported simulators see :ref:`_simulator-support`.
 
-Windows Installation
---------------------
+Install and ensure that simulator executable is in the ``PATH``.   
+.. _installation_via_pip:
```
Comment (cmarqu) [inline documentation/source/quickstart.rst:70]:
```suggestion
.. _installation_via_pip::
```

File: documentation/source/quickstart.rst [hunk 4]
```diff
@@ -5,8 +5,8 @@ Quickstart Guide
 Installing cocotb
 =================
 
-Pre-requisites
---------------
+Pre-requisites Installation
```
Comment (cmarqu) [inline documentation/source/quickstart.rst:8]:
```suggestion
Installation of Pre-requisites
```

**Comment by cmarqu** [comment]
Stepping a bit back, I would prefer if this moved out of ``quickstart.rst`` and into an ``install.rst`` in this PR already.

The main heading in ``install.rst`` would be "Installation",
and the next lines would say e.g.
```
Using cocotb requires installation of pre-requisites and installation of cocotb itself.
In this tutorial we are assuming that you already have a :ref:`supported simulator<simulator-support>` in ``PATH``.
```
so that we have that out of our way.

Then, have subsections for Windows, Linux, macOS, where we explain conda for Windows, native installation for Linux, brew for macOS.
Point to that Tier 2 wiki page (I'm not so happy with that name TBH) at the end of each OS subsection where we can put up more complicated instructions.

This is so that e.g. a Windows user can (mentally or maybe even in Sphinx) hide all information that does not have "Windows" written on it.

cocotb PIP install instructions... The new "Key Topics" section in the docs is the place to explain installation from GitHub, together with info on how to test a PR of someone else. This could be in a file ``install_devel.rst``.

(Enough for tonight :))

**Comment by ktbarrett** [comment]
Could you add one more section in your document about installing using only a local clone of cocotb, and mention that you may need `--no-use-pep517`?

**Comment by themperek** [comment]
@cmarqu For me definitely easier would be to merge as is and you can later restructure as you like. Or just take those changes and incorporate them in your PR.

**Comment by themperek** [comment]
@ktbarrett 
>Could you add one more section in your document about installing using only a local clone of cocotb, and mention that you may need --no-use-pep517?

Separate PR? We have a link/note for local install.

File: documentation/source/quickstart.rst [hunk 5]
```diff
@@ -18,104 +18,92 @@ Cocotb has the following requirements:
 
 .. versionchanged:: 1.4 Dropped Python 2 support
 
-.. _installation_via_pip:
+.. note:: In order to use a 32bit simulator you need to use a 32bit version of Python.
```
Comment (eric-wieser) [inline documentation/source/quickstart.rst:21]:
```suggestion
.. note:: In order to use a 32-bit simulator you need to use a 32-bit version of Python.
```

File: documentation/source/quickstart.rst [hunk 6]
```diff
@@ -18,104 +18,92 @@ Cocotb has the following requirements:
...
+.. code-block:: bash
 
-Native Linux Installation
--------------------------
+    sudo yum install make gcc gcc-c++ libstdc++-devel swig python3 python3-devel python3-pip
 
-The following instructions will allow building of the cocotb libraries
-for use with a 64-bit native simulator.
+macOS Packages
+^^^^^^^^^^^^^^
```
Comment (eric-wieser) [inline documentation/source/quickstart.rst:53]: For consistency with the other headings:
```suggestion
Installation for macOS
^^^^^^^^^^^^^^^^^^^^^^
```

File: documentation/source/quickstart.rst [hunk 7]
```diff
@@ -18,104 +18,92 @@ Cocotb has the following requirements:
...
 
-To install the development version of cocotb:
+Native Installation for Debian/Ubuntu-based Systems
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 .. code-block:: bash
 
-    git clone https://github.com/cocotb/cocotb
-    pip3 install -e ./cocotb
+    sudo apt-get install make gcc g++ swig python3 python3-dev python3-pip
```
Comment (eric-wieser) [inline documentation/source/quickstart.rst:43]:
```suggestion
    sudo apt-get install make gcc g++ python3 python3-dev python3-pip
    sudo apt-get install swig  # optional, needed for one of the examples
```

File: documentation/source/quickstart.rst [hunk 8]
```diff
@@ -18,104 +18,92 @@ Cocotb has the following requirements:
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-    After installation, you should be able to execute ``cocotb-config``.
-    If it is not found, you need to append its location to the ``PATH`` environment variable.
-    This may happen when you use the ``--user`` option to ``pip``, in which case the location is documented :ref:`here <python:inst-alt-install-user>`.
+.. code-block:: bash
 
-Native Linux Installation
--------------------------
+    sudo yum install make gcc gcc-c++ libstdc++-devel swig python3 python3-devel python3-pip
```
Comment (eric-wieser) [inline documentation/source/quickstart.rst:50]:
```suggestion
    sudo yum install make gcc gcc-c++ libstdc++-devel swig python3 python3-devel python3-pip
    sudo yum install swig  # optional, needed for one of the examples
```

File: documentation/source/quickstart.rst [hunk 9]
```diff
@@ -18,104 +18,92 @@ Cocotb has the following requirements:
...
 
-    sudo apt-get install git make gcc g++ swig python-dev
+    brew install python icarus-verilog gtkwave
 
-Red Hat-based
-^^^^^^^^^^^^^
+.. seealso::
 
-.. code-block:: bash
+   For more installation options (also for 32-bit simulators) please see `our Wiki <https://github.com/cocotb/cocotb/wiki/Tier-2-Setup-Instructions>`_.
```
Comment (eric-wieser) [inline documentation/source/quickstart.rst:66]: I think this should be below the `Simulator Installation` heading

File: documentation/source/quickstart.rst [hunk 10]
```diff
@@ -18,104 +18,92 @@ Cocotb has the following requirements:
...
-"Mark for installation"
+    Ensure that path to the simulator executable is appended to the ``PATH`` environment variable.
 
-.. code-block:: bash
 
-    Basic Installation
-      -- mingw-developer-tools
-      -- mingw32-base
-      -- mingw32-gcc-g++
-      -- msys-base
```
Comment (eric-wieser) [inline documentation/source/quickstart.rst:92]: Are we sure we want to remove these instructions? Maybe worth putting them on the wiki first, so they don't get lost completely.

File: documentation/source/quickstart.rst [hunk 11]
```diff
@@ -18,104 +18,92 @@ Cocotb has the following requirements:
...
-Python can be downloaded from https://www.python.org/downloads/windows/.
-Run the installer and download to your chosen location.
+Cocotb can be installed by running
 
-It is beneficial to add the path to Python to the Windows system ``PATH`` variable
-so it can be used easily from inside Msys.
+.. code-block:: bash
 
-Once inside the Msys shell commands as given here will work as expected.
+    pip install cocotb  # use `pip3` in case multiple of Python/pip version installed
```
Comment (eric-wieser) [inline documentation/source/quickstart.rst:90]:
```suggestion
    pip install cocotb  # use `pip3` in case multiple versions of Python or pip are installed
```

File: documentation/source/quickstart.rst [hunk 12]
```diff
@@ -18,104 +18,92 @@ Cocotb has the following requirements:
...
 
-macOS Packages
---------------
+.. seealso::
 
-You need a few packages installed to get cocotb running on macOS.
-Installing a package manager really helps things out here.
+    For user local installation follow the `pip User Guide <https://pip.pypa.io/en/stable/user_guide/#user-installs/>`_.
```
Comment (eric-wieser) [inline documentation/source/quickstart.rst:94]:
```suggestion
    For user-local installation, follow the `pip User Guide <https://pip.pypa.io/en/stable/user_guide/#user-installs/>`_.
```

File: documentation/source/quickstart.rst [hunk 13]
```diff
@@ -18,104 +18,92 @@ Cocotb has the following requirements:
...
 
-`Brew <https://brew.sh/>`_ seems to be the most popular, so we'll assume you have that installed.
+To install the development version of cocotb:
 
 .. code-block:: bash
 
-    brew install python icarus-verilog gtkwave
+    pip install https://github.com/cocotb/cocotb/archive/master.zip  # use `pip3` in case multiple of Python/pip version installed
```
Comment (eric-wieser) [inline documentation/source/quickstart.rst:100]:
```suggestion
    pip install https://github.com/cocotb/cocotb/archive/master.zip  # use `pip3` in case multiple versions of Python or pip are installed
```

File: documentation/source/simulator_support.rst [hunk 14]
```diff
@@ -124,14 +129,19 @@ See :ref:`sim-modelsim`.
 Mentor ModelSim
 ===============
 
-Any ModelSim PE or ModelSim PE derivative (like ModelSim Microsemi, Intel, Lattice Edition) does not support the VHDL FLI feature.
-If you try to run with FLI enabled, you will see a ``vsim-FLI-3155`` error:
+.. note::
+
+   In order to use FLI (for VHDL), a ``vdbg`` executable from the simulator installation directory needs to be available during cocotb installation.
```
Comment (eric-wieser) [inline documentation/source/simulator_support.rst:134]:
```suggestion
   In order to use FLI (for VHDL), a ``vdbg`` executable from the simulator installation directory needs to be available on the ``PATH`` during cocotb installation.
```

File: documentation/source/quickstart.rst (hunk 10 above)
Comment (themperek) [reply documentation/source/quickstart.rst:92]: I think yes. On the wiki, I can add msys2 instructions.

File: documentation/source/quickstart.rst (hunk 11 above)
Comment (themperek) [reply documentation/source/quickstart.rst:92]: Maybe even better would be ``use `pip3` if available`` ?

Comment (themperek) [reply documentation/source/quickstart.rst:92]: Or better opposite 
```bash
pip3 install cocotb  # use `pip` if `pip3` not available
```

Comment (eric-wieser) [reply documentation/source/quickstart.rst:92]: How about we just use
```
pip install cocotb
```
but add a note before the first time we mention `pip` like
```
.. warning::
    ``pip`` may belong to a different python installation to what you expect.
    Use ``pip -V`` to check.
    If this shows Python 2.7, use ``pip3`` or ``python3 -m pip`` in place of ``pip`` in the commands below.
```

Comment (themperek) [reply documentation/source/quickstart.rst:92]: Sound OK. Yo sure **before** or rather close to `cocotb-config` note?

Comment (eric-wieser) [reply documentation/source/quickstart.rst:92]: You're right, your suggested placement makes more sense.

Comment (themperek) [reply documentation/source/quickstart.rst:92]: :+1: If we see issues we move up ;-)

**Comment by imphil** [comment]
Thanks for updating the quickstart. I just had a look at some numbers for the quickstart page, and it turns out (not very surprisingly) it is the most viewed page in our documentation by far.
Since this page is also the entry point for new users we should invest a bit into making it as tailored to our users as possible and reduce the risk of users getting stuck in the middle of the process.
To achieve that I'd use the following guideline:

- Make the steps very actionable. Step-by-step instructions which ideally can be copy-pasted without any thinking.
- Make the steps robust. If there's a "theoretically better/faster/cleaner" and a more robust version, choose the more robust one.
- Make the steps linear, avoid "branching" documentation more than once. For example, don't have a pre-requisites section where the user needs to pick their OS, and then have a "main install" section where the user needs to do the same.
- Make expectations clear, and provide checkpoints. If a command has output, show at least part of the output. If there's a file which should be present after a step, tell the user.
- Order instructions so that the most commonly used ones are at the top.

To answer the ordering question, I had a look at some numbers:

```
Site views for the quickstart page
50 % Windows (mostly Win 10)
22 % Linux
18 % Mac


cocotb users from the current survey:

Ubuntu 18.04 (38%)
Ubuntu 16.04 (10%)
Debian (no version) 11 %
RHEL6: 21%
RHEL7: 23%
RHEL8: 7%
Arch: 9%

Windows 7/8: 8%
Windows 10: 13%
Windows 10/WSL: 7%
```

Applying this to the quickstart page, I envision a structure like this:

- Generic quickstart for experienced users: list pre-requisites, and "pip install cocotb"
- Step-by-step install instructions for specific (groups) of operating systems. Probably something along those lines
  - Ubuntu/Debian
  - Windows 7/8/10 (as long as they are identical)
  - RHEL (we probably need to figure out if we want to go the conda route there, or use SCLs.)

  Each of these sections should be a step-by-step installation instruction starting with the installation of a supported version of Python (e.g. via Conda or if possible via the system package manager), and end with a check that cocotb now works (e.g. by calling cocotb-config).

On the quickstart page do not include steps how to install a development version of cocotb. This information should be on a separate page in the developer section of the docs (you can link to it).

To keep this PR manageable, I'd propose to start with the page structure and the generic bits, we can then fill in piece-by-piece specialized blocks in subsequent PRs.

**Comment by eric-wieser** [comment]
Is it worth putting this PR in as-is, and addressing your suggestions in a new PR?

**Comment by themperek** [comment]
@imphil This is perfectly fine with me but this PR mostly updates information that was wrong/missing. I think it is above the scope of this PR and my capability to do restructuring right. I am willing to help for improvements/restructuring.

**Comment by imphil** [comment]
Yes, fine either way.

**Comment by eric-wieser** [comment]
> Sound OK. Are you sure **before** or rather close to `cocotb-config` note?

@themperek, did you get around to making this change?

File: documentation/source/quickstart.rst (hunk 9 above)
Comment (themperek) [reply documentation/source/quickstart.rst]: I have changed this a bit to remove `simulators`  since this was the intention. Feel free to change.

File: .vscode/settings.json [hunk 15]
```diff
@@ -0,0 +1,4 @@
+{
+    "restructuredtext.confPath": "${workspaceFolder}\\documentation\\source",
+    "python.pythonPath": "C:\\Users\\Tomek\\Miniconda3\\python.exe"
+}
```
Comment (imphil) [inline .vscode/settings.json:4]: That should not be part of this commit.

Comment (themperek) [reply .vscode/settings.json]: Sorry.

File: documentation/source/quickstart.rst (hunk 10 above)
Comment (themperek) [reply documentation/source/quickstart.rst]: Added instruction for MSYS/MINGW here: https://github.com/cocotb/cocotb/wiki/Tier-2-Setup-Instructions#windwows-installation-using-msys2

**Comment by themperek** [comment]
Thank you all for help and sorry for all the mistakes.

**Comment by imphil** [comment]
No worries, thanks for your persistence!
